include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+
+ private:
+  inline void collect_pending_unsafe(cudaStream_t ctx) {
+    auto it = std::remove_if(
+        pending_.begin(), pending_.end(),
+        [this, ctx](const std::pair<alloc_type*, cudaEvent_t>& pair) {
+          const cudaError_t event_state = cudaEventQuery(pair.second);
+          switch (event_state) {
+            case cudaSuccess:
+              // Stock buffers and destroy those that
+              // are no longer needed.
+              if (stock_.size() < options.max_stock) {
+                stock_.emplace_back(pair.first);
+              } else {
+                Allocator::free(pair.first, ctx);
+              }
+              ready_events_.emplace_back(pair.second);
+              return true;
+            case cudaErrorNotReady:
+              return false;
+            default:
+              CUDA_CHECK(event_state);
+              return false;
+          }
+        });
+    pending_.erase(it, pending_.end());
+  }
+
+  template <class Iterator>
+  inline void get_raw(Iterator first, Iterator const last, cudaStream_t ctx) {
+    // Get pre-allocated buffers if stock available.
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+
+      for (; first != last; ++first) {
+        // If no buffers available, try to make some available.
+        if (stock_.empty()) {
+          collect_pending_unsafe(ctx);
+          if (stock_.empty()) {
+            // No buffers available.
+            break;
+          }
+        }
+
+        // Just take the next available buffer.
+        *first = stock_.back();
+        stock_.pop_back();
+      }
+    }
+
+    // Forge new buffers until request can be filled.
+    for (; first != last; ++first) {
+      *first = Allocator::alloc(options.buffer_size, ctx);
+    }
+  }
+
+  template <class Iterator>
+  inline void put_raw(Iterator first, Iterator const last, cudaStream_t ctx) {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (ctx) {
+      for (; first != last; ++first) {
+        // Spin lock if too many pending buffers (i.e., let CPU wait for GPU).
+        while (ready_events_.empty()) {
+          collect_pending_unsafe(ctx);
+          if (!ready_events_.empty()) {
+            break;
+          }
+          std::this_thread::yield();
+        }
+
+        // Queue buffer.
+        cudaEvent_t ready_event = ready_events_.back();
+        ready_events_.pop_back();
+        CUDA_CHECK(cudaEventRecord(ready_event, ctx));
+        pending_.emplace_back(*first, ready_event);
+      }
+    } else {
+      // Without stream context, we must force a hard sync with the GPU.
+      CUDA_CHECK(cudaDeviceSynchronize());
+
+      for (; first != last; ++first) {
+        // Stock buffers and destroy those that are no longer needed.
+        if (stock_.size() < options.max_stock) {
+          stock_.emplace_back(*first);
+        } else {
+          Allocator::free(*first);
+        }
+      }
+    }
+  }
+
```

Comment:
` private:` would be more clear, for the previous private method is too long.

Reply:
There is a `private:` there above the private functions.